### PR TITLE
feat(srp): add a quiz to the SRP reveal

### LINKED ISSUE #17557: Add friction to SRP reveal
# Description

Following the implementation on mobile, it's necessary to develop the "Reveal SRP Quiz" on extension.

> _Through user testing it’s been shown that users are ignoring the warnings around SRP safety when they go to reveal their SRP. As such, it’s assumed that users ignoring these warnings are contributing in some way to phishing fund loss incidents._

Design: https://www.figma.com/file/4LQje3epQz8g4xO8fRdauv/SRP-Reveal-Warning?node-id=4119%3A13076&t=NlcYAWjU62vsTcXf-1

# Technical Details

- All the views must use the specified DS components
- All the code has to be written in TypeScript and any new component must be a functional component

# Acceptance Criteria

- Epic metrics in place
- Multiple languages supported
- The quiz is developed according the defined UI/UX using the current extension DS

# References

- [Epic file (Mobile focused)](https://docs.google.com/document/d/1kzI0BiBT9-6LuEp19lfZu0RV7wnXLOX24zg0r4uxK0A/edit#)

### PR DESCRIPTION
**Pull request by HowardBraham**
## Explanation

* Fixes #17557
* Fixes https://github.com/MetaMask/MetaMask-planning/issues/597
* Fixes https://github.com/MetaMask/MetaMask-planning/issues/268

[Figma file](https://www.figma.com/file/4LQje3epQz8g4xO8fRdauv/SRP-Reveal-Warning)
[Google Doc](https://docs.google.com/document/d/1kzI0BiBT9-6LuEp19lfZu0RV7wnXLOX24zg0r4uxK0A)

Uses the DS component `ui/components/component-library/modal`

**100% test coverage for added files and for `/ui/pages/settings/security-tab` folder**

### After
image

## Manual Testing Steps

- Go to Settings -> Security & privacy
- Click Reveal Secret Recovery Phrase
- Get the questions right, get the questions wrong, make sure it goes to the right place
- If you have access to MixPanel, make sure it fires the right events at the right time

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/app/srp-quiz-popover/QuizContent/QuizContent.tsx
Comment (darkwing) [inline ui/components/app/srp-quiz-popover/QuizContent/QuizContent.tsx:34]: We should have a localized `alt` attribute here.

File: ui/components/app/srp-quiz-popover/QuizContent/QuizContent.tsx
Comment (darkwing) [inline ui/components/app/srp-quiz-popover/QuizContent/QuizContent.tsx:39]: `TextColor` has an `inherit` value

File: ui/components/app/srp-quiz-popover/SRPQuiz/SRPQuiz.tsx
Comment (darkwing) [inline ui/components/app/srp-quiz-popover/SRPQuiz/SRPQuiz.tsx:52]: We should probably add this URL to `ui/helpers/constants/zendesk-url.js` and reference it here.

File: ui/components/app/srp-quiz-popover/QuizContent/QuizContent.tsx
Comment (HowardBraham) [reply ui/components/app/srp-quiz-popover/QuizContent/QuizContent.tsx:39]: Whoops, "inherit" was a confusing word to write in the comment.  It's not inheriting the color of its parent, but the color of its sibling.  So maybe I should change the comment to say "copy."

File: ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx
Comment (montelaidev) [inline ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx:50]:
```suggestion
          variant={TextVariant.bodyMd}
```
moreContent looks like it is supposed to be TextVariant.bodyMd

File: ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx
Comment (montelaidev) [inline ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx:41]:
```suggestion
        variant={TextVariant.bodyLg}
```
This looks like it should be using the TextVariant.headingLg

**Comment by HowardBraham** [comment]
@brad-decker I had to add `@types/react-router-dom` because there was no Typescript file that was using `react-router-dom` before.  I guess SocketSecurity is complaining that it might be unmaintained.  What's the right procedure here?

**Comment by HowardBraham** [comment]
@brad-decker @danjm  I had to add @types/react-router-dom because there was no Typescript file that was using react-router-dom before. I guess SocketSecurity is complaining that it might be unmaintained. What's the right procedure here?

**Review by georgewrmarshall — COMMENTED** [review]
This is looking great!! I checked out your branch and played around with unifying some of the spacing and bit more and created a PR against this one https://github.com/MetaMask/metamask-extension/pull/19367

**Review by georgewrmarshall — COMMENTED** [review]
Thanks for being so responsive to my suggestions. I've left some more to fix some of the oversights on my part apologies! Below is screen recording with suggestions

https://github.com/MetaMask/metamask-extension/assets/8112138/d1498d09-c2af-4e84-9f3c-b161ffad30ca

One other note is that could we maybe update the Settings page section so we don't have such a long character length for the Button and maybe match the other sections a bit more. I do like the simplicity of it currently but we could add some more information? Below is a rough mock up and will need some design input from @holantonela and @coreyjanssen 

## Current
Screenshot 2023-06-01 at 2 11 11 PM


## Proposed
Screenshot 2023-06-01 at 2 09 58 PM

File: app/images/reveal-srp.png
Comment (HowardBraham) [reply app/images/reveal-srp.png:1]: Oh yes, I did notice that the PNG went from 19kb to 24kb while also getting smaller

Comment (georgewrmarshall) [reply app/images/reveal-srp.png:1]: Yup totally my bad 🤦 sorry!

Comment (holantonela) [reply app/images/reveal-srp.png:1]: Thanks for this suggestion @georgewrmarshall! I think is pretty nice that we are anticipating the next flow. We discussed it before. I'm OK implementing it. We may want to have the OK from @coreyjanssen and @AlexJupiter to move forward with this. We also need to match this on Mobile.

Comment (plasmacorral) [reply app/images/reveal-srp.png:1]: I am a bit suprised to see this proposal to present a CTA with the abbreviation "SRP".  While I do recognize the impact of long strings to translate, I thought we previously had a mind to NOT abbreviate Secret Recovery Phrase.  Has that that changed?  Where do we declare this abbreviation for the user by spelling out SRP and following it with "(SRP)" to indication the equivocation?

**Comment by plasmacorral** [comment]
> Thanks for being so responsive to my suggestions. I've left some more to fix some of the oversights on my part apologies! Below is screen recording with suggestions
> after.mov
> 
> One other note is that could we maybe update the Settings page section so we don't have such a long character length for the Button and maybe match the other sections a bit more. I do like the simplicity of it currently but we could add some more information? Below is a rough mock up and will need some design input from @holantonela and @coreyjanssen
> ## Current
> Screenshot 2023-06-01 at 2 11 11 PM
> ## Proposed
> 
> Screenshot 2023-06-01 at 2 09 58 PM

@georgewrmarshall I am a bit suprised to see this proposal to present a CTA with the abbreviation "SRP".  While I do recognize the impact of long strings to translate, I thought we previously had a mind to NOT abbreviate Secret Recovery Phrase.  Has that that changed?  Where do we declare this abbreviation for the user by spelling out "Secret Recovery Phrase" and following it with "(SRP)" to indicate the equivocation?

**Comment by coreyjanssen** [comment]
@plasmacorral agreed re: spelling out Secret Recovery Phrase! In addition, there are other content issues that should be addressed before launching this.

**Comment by plasmacorral** [comment]
@HowardBraham can we revert to what was initially approved for this effort, and pursue some of George's proposed changes after they are further refined for a phase 2?

**Comment by coreyjanssen** [comment]
Modified copy here!

**Reveal Secret Recovery Phrase**
**[body]** To keep your wallet safe, we'll ask a couple questions before revealing your Secret Recovery Phrase. 
**[button]** Get started

**Comment by HowardBraham** [comment]
> @HowardBraham can we revert to what was initially approved for this effort

I don't actually think there was anything that was initially approved, but I did implement and push the changes from George and Corey.

**Comment by brad-decker** [comment]
@socket-security ignore-all

**Comment by brad-decker** [comment]
@SocketSecurity ignore-all

File: app/_locales/en/messages.json
Comment (georgewrmarshall) [inline app/_locales/en/messages.json:3594]: I know it had a period in the comment but not sure if it was a mistake. I believe the guideline is to not use periods unless it's part of a multiple sentence description https://www.notion.so/MetaMask-Content-Design-Style-Guide-7d0b7579786540afbe02780ebb203254?pvs=4

Screenshot 2023-06-05 at 3 25 24 PM

best check with @coreyjanssen

**Review by georgewrmarshall — COMMENTED** [review]
UI looks good!! Tested both the flow in popup and expanded view.

Only[ issue is that the heading is cut off](https://github.com/MetaMask/metamask-extension/issues/19458) which is not apart of the scope of this ticket and should be resolved in another PR cc @darkwing 

 This is oversight on my part as I did not provide @coreyjanssen with the entire flow in my proposal and I don't want to block this any further but there is almost a redundant screen. We are using "Get Started" as a CTA twice. It's almost as if the content from the first screen of the Modal could be used in the settings page? cc @holantonela 


https://github.com/MetaMask/metamask-extension/assets/8112138/a70ae1b9-505d-40a5-8e2c-91ef7e3f2289

File: ui/pages/settings/security-tab/security-tab.component.js
Comment (montelaidev) [inline ui/pages/settings/security-tab/security-tab.component.js:46]: should this stay as required?

Comment (HowardBraham) [reply ui/pages/settings/security-tab/security-tab.component.js:46]: It's technically not required, since in `security-tab.container.js` it has this default value:
`featureFlags: { showIncomingTransactions } = {},`

I changed this `isRequired` so I could write a new test and get 100% test coverage on `security-tab.test.js`

Now one could argue that's maybe the wrong decision... that the right thing would be to remove the default value and leave it as `isRequired`.  So I guess let's discuss that.

**Review by montelaidev — DISMISSED** [review]

**Review by georgewrmarshall — COMMENTED** [review]
LGTM. Only thing I can see is that the image needs the background removed so it looks good in dark mode. Image below

Screenshot 2023-06-06 at 1 34 06 PMScreenshot 2023-06-06 at 1 33 30 PM


reveal-srp

**Comment by HowardBraham** [comment]
> LGTM. Only thing I can see is that the image needs the background removed so it looks good in dark mode. Image below

The [original file on the first commit](https://github.com/MetaMask/metamask-extension/blob/1fe0cb3c95fafe67f8d8dad2873160807692ba3c/app/images/reveal-srp.png) had a transparent background.  Was that the right image after all? 🙂

**Review by georgewrmarshall — DISMISSED** [review]
LGTM! Really nice work 💯

**Review by montelaidev — DISMISSED** [review]

**Review by montelaidev — DISMISSED** [review]

**Review by georgewrmarshall — DISMISSED** [review]
LGTM! UI is looking 🔥🔥🔥
- Pulled branch and tested quiz locally through to revealing my SRP

**Review by montelaidev — DISMISSED** [review]

**Review by plasmacorral — CHANGES_REQUESTED** [review]
Translation strings are not currently included and are outlined as required in acceptance criteria 9. Once the translation strings are merged, we can move forward.

**Comment by plasmacorral** [comment]
Have reviewed the quiz in the following languages, with mobile translations as the source:

(de) German
(el) Greek
(es) Spanish
(fr) French
(hi) Hindi
(id) Indonesian
(ja) Japanese
(ko) Korean
(pt) Portuguese
(ru) Russian
(tl) Tagalog
(tr) Turkish
(vi) Vietnamese
(zh) Chinese

In both regular and expanded view for completeness.  Using Chrome Version 114.0.5735.106 (Official Build) (arm64) on Ventura 13.4.  

srpSecurityQuizImgAlt will be added once translations are available.

**Comment by plasmacorral** [comment]
Did have a strange observation while testing Portuguese where the Get started page was not presented after tapping the reveal SRP CTA from Security and Privacy, and the first question was presented without explanation.  This was only observed one time even though multiple languages were verified.  On retry, get started title, image and description were displayed as expected.

Using Chrome Version 114.0.5735.106 (Official Build) (arm64) on Ventura 13.4.
See video: https://recordit.co/TUFFXEFUcT

**Comment by plasmacorral** [comment]
While testing in French, the Get started page was presented without a title initially.  This was only observed one time even though multiple languages were verified. On retry, get started title, image and description were displayed as expected.

**Review by plasmacorral — DISMISSED** [review]

**Review by montelaidev — DISMISSED** [review]

File: ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx
Comment (georgewrmarshall) [inline ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx:13]: If you rebase the `Box` component in TypeScript will be available from the `component-library` it would be a 2 line update as the component API remains the same

```suggestion
import { Button, Text, Box } from '../../../component-library';
```

**Review by georgewrmarshall — COMMENTED** [review]
I've found a bug when using the keyboard to navigate through the quiz. The `title` get's reset to an empty string when tabbing through. It seems tabbing is causing a re-render and because `title` is not a state variable it is resetting to an empty string. I'm not sure what is causing the re-render but it could be the `ModaFocus` component that's used to improve the accessibility of the `Modal`. I know you've stated the reason for `title` not being a state variable in the code comments but I would recommend updating it to one and we could investigate the re-render in a separate issue?

https://github.com/MetaMask/metamask-extension/assets/8112138/78c0cd11-22c8-4052-b55c-c319425c454b

**Comment by HowardBraham** [comment]
> I've found a bug when using the keyboard to navigate through the quiz. The `title` get's reset to an empty string when tabbing through. It seems tabbing is causing a re-render and because `title` is not a state variable it is resetting to an empty string. I'm not sure what is causing the re-render but it could be the `ModaFocus` component that's used to improve the accessibility of the `Modal`. I know you've stated the reason for `title` not being a state variable in the code comments but I would recommend updating it to one and we could investigate the re-render in a separate issue?

This turned out to be quite complicated, lemme break it down into steps:
1. The re-render was caused by the modal's onClose changing, because security-tab.component.js had `onClose={() => this.setState({ srpQuizPopoverVisible: false })}`, which technically passes a different function with every re-render.  I made onClose not change, and also changed the word `Popover` to `Modal` (whoops, we all missed this before).
2. It was still re-rendering when I hit the Tab key, because `Route` was passing a different `props.match`.  I tried to stop this, but I'm not sure I can.
3. If there was ever a re-render without `stage` changing, `title` would not be computed.  I guess I thought this never happened, but the Tab key does it.  I had been a little too overeager with the `useMemo()`, so I made `quizContent` get computed on every render, and now it works.

**Review by georgewrmarshall — APPROVED** [review]
Oh wow that does sound complicated. Really nice work getting to the bottom of it. Thanks for the clear explanation and putting in the effort to resolve. I've run through the quiz again locally, keyboard navigation is working perfectly and it's looking great!! LGTM!

**Review by montelaidev — APPROVED** [review]

**Comment by plasmacorral** [comment]
Observed an error in Firefox 114, on ventura 13.4 with commit d6771db503600db9ec8970ee8161cc71e4e2c1a7 built locally.

error

See video, errors seems to occur when hold to reveal SRP is closed: https://recordit.co/AegOKk9Xtt

See the same steps on production v10.31.1 do not produce this error: http://recordit.co/HLO7VchHjB

**Pull request merged**
